### PR TITLE
3600 - 3

### PR DESCRIPTION
**Pull request by sametsJR**
## Номер
3600

## Номер задания
3

## Ссылка на видео с демо
https://youtu.be/Jg4AL7wVm4c

## Комментарии
Done task 3

### CODE REVIEW

File: 3600/3/Gemfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+source 'https://rubygems.org'
+
+gem 'dotenv'
```
Comment (Moleved) [inline 3600/3/Gemfile]: If you do
```ruby
gem 'dotenv', require: 'dotenv/load'
```
`Bundler.require` will require and load dotenv, so won't be necessary `Dotenv.load`

File: 3600/3/app/models/users.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+class User < ActiveRecord::Base
```
Comment (Moleved) [inline 3600/3/app/models/users.rb]: Add validations at least for user

Comment (Moleved) [reply 3600/3/app/models/users.rb]: When be done also add some of the validations to the database by migration

File: 3600/3/app/views/layout.erb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+<!DOCTYPE html>
```
Comment (Moleved) [inline 3600/3/app/views/layout.erb]: You have layout file but you don't use it. Don't do so

File: 3600/3/app/controllers/app_controller.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+class AppController < Sinatra::Base
+  Dotenv.load
+  configure do
+    enable :sessions
+    set :public_folder, File.dirname(__FILE__) + '/public'
+    set :session_secret, ENV['key']
+  end
+
+  use UserController
+  use RestaurantController
```
Comment (Moleved) [inline 3600/3/app/controllers/app_controller.rb]: Put this two lines to `config.ru` file

File: 3600/3/app/controllers/restaurant_controller.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+require_relative '../../helpers/comment_helper.rb'
+
+class RestaurantController < Sinatra::Base
```
Comment (Moleved) [inline 3600/3/app/controllers/restaurant_controller.rb]: Inherit it from AppController

File: 3600/3/app/controllers/user_controller.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+require_relative '../../helpers/user_helper'
+require 'digest'
+
+class UserController < Sinatra::Base
+  set views: proc { File.join(root, '../views/') }
```
Comment (Moleved) [inline 3600/3/app/controllers/user_controller.rb]: Move to AppController

File: 3600/3/app/controllers/app_controller.rb (hunk 4 above)
Comment (Moleved) [reply 3600/3/app/controllers/app_controller.rb]: Before `run AppController`

File: 3600/3/Gemfile [hunk 7]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
```
[the remark this answers is not in the document]
Comment (sametsJR) [reply 3600/3/Gemfile]: Ok

File: 3600/3/Gemfile (hunk 1 above)
Comment (sametsJR) [reply 3600/3/Gemfile]: Removed

File: 3600/3/app/views/layout.erb (hunk 3 above)
Comment (sametsJR) [reply 3600/3/app/views/layout.erb]: Removed

File: 3600/3/app/controllers/app_controller.rb (hunk 4 above)
Comment (sametsJR) [reply 3600/3/app/controllers/app_controller.rb]: Done

File: 3600/3/app/controllers/user_controller.rb (hunk 6 above)
Comment (sametsJR) [reply 3600/3/app/controllers/user_controller.rb]: Moved

File: 3600/3/app/controllers/restaurant_controller.rb (hunk 5 above)
Comment (sametsJR) [reply 3600/3/app/controllers/restaurant_controller.rb]: Done

File: 3600/3/app/controllers/app_controller.rb (hunk 4 above)
Comment (sametsJR) [reply 3600/3/app/controllers/app_controller.rb]: Done

File: 3600/3/app/views/layout.erb (hunk 3 above)
Comment (sametsJR) [reply 3600/3/app/views/layout.erb]: Without layout file:
Screenshot from 2019-08-02 15-54-03
Screenshot from 2019-08-02 15-55-17

File: 3600/3/Gemfile (hunk 7 above)
[the remark this answers is not in the document]
Comment (sametsJR) [reply 3600/3/Gemfile]: Removed

File: 3600/3/Gemfile (hunk 1 above)
Comment (sametsJR) [reply 3600/3/Gemfile]: Dotenv.load deleted

File: 3600/3/app/controllers/app_controller.rb (hunk 4 above)
Comment (sametsJR) [reply 3600/3/app/controllers/app_controller.rb]: Done

File: 3600/3/app/controllers/restaurant_controller.rb (hunk 5 above)
Comment (sametsJR) [reply 3600/3/app/controllers/restaurant_controller.rb]: Inherited

File: 3600/3/app/controllers/user_controller.rb (hunk 6 above)
Comment (sametsJR) [reply 3600/3/app/controllers/user_controller.rb]: Moved
Screenshot from 2019-08-02 22-51-25

File: 3600/3/app/controllers/restaurant_controller.rb (hunk 5 above)
Comment (sametsJR) [reply 3600/3/app/controllers/restaurant_controller.rb]: Screenshot from 2019-08-02 22-52-38

File: 3600/3/app/controllers/app_controller.rb (hunk 4 above)
Comment (sametsJR) [reply 3600/3/app/controllers/app_controller.rb]: Screenshot from 2019-08-02 22-53-25

File: 3600/3/Gemfile (hunk 1 above)
Comment (sametsJR) [reply 3600/3/Gemfile]: And I don't understand where you got this from: require: 'dotenv/load'

File: 3600/3/app/models/users.rb (hunk 2 above)
Comment (sametsJR) [reply 3600/3/app/models/users.rb]: Screenshot from 2019-08-02 23-13-39

File: 3600/3/Gemfile (hunk 1 above)
Comment (Moleved) [reply 3600/3/Gemfile]: From documentation of bundler

Comment (Moleved) [reply 3600/3/Gemfile]: WIthout `require: 'dotenv/load'` you `.env` file won't be loaded

Comment (sametsJR) [reply 3600/3/Gemfile]: Those I just have to add require: 'dotenv/load' to gem 'dotenv'?

File: 3600/3/app/views/layout.erb (hunk 3 above)
Comment (Moleved) [reply 3600/3/app/views/layout.erb]: To use layout
```ruby
erb :index, layout: :layout
```

File: 3600/3/.gitignore [hunk 8]
```diff
@@ -0,0 +1 @@
+/db
```
Comment (Moleved) [inline 3600/3/.gitignore]: Replace with `db/*.sqlite3`

File: 3600/3/.gitignore [hunk 9]
```diff
@@ -0,0 +1 @@
+db/*.sqlite3
```
Comment (Moleved) [inline 3600/3/.gitignore]: My mistake `db/*.sqlite` to exclude database file

File: 3600/3/app/controllers/app_controller.rb [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+  configure do
+    register Sinatra::ActiveRecordExtension
+    register Sinatra::Flash
+    enable :sessions
+    set :public_folder, File.dirname(__FILE__) + '/public'
+    set :session_secret, ENV['key']
+  end
+
+  get '/' do
+    erb :index, layout: :layout
```
Comment (Moleved) [inline 3600/3/app/controllers/app_controller.rb]: Do you have `index` file?

File: 3600/3/app/controllers/restaurant_controller.rb [hunk 11]
```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+require_relative '../../helpers/comment_helper.rb'
```
Comment (Moleved) [inline 3600/3/app/controllers/restaurant_controller.rb]: Should be already required by `require_all 'app'`

File: 3600/3/helpers/comment_helper.rb [hunk 12]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+module CommentsHelper
+  def add_new_comment
+    @user = User.find(session[:user_id])
+    @comment = @user.comments.create(grade: params[:grade].to_i, text: params[:text])
+    @comment.restaurant_id = @restaurant.id
+    @comment.save
```
Comment (Moleved) [inline 3600/3/helpers/comment_helper.rb]: Take a look at the source code of `create` method https://apidock.com/rails/ActiveRecord/Base/create/class

File: 3600/3/helpers/comment_helper.rb [hunk 13]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+module CommentsHelper
+  def add_new_comment
+    @user = User.find(session[:user_id])
+    @comment = @user.comments.create(grade: params[:grade].to_i, text: params[:text])
```
Comment (Moleved) [inline 3600/3/helpers/comment_helper.rb]:
```ruby
@user.comments.create(grade: params[:grade].to_i, text: params[:text], restaurant_id: @restaurant.id)
```
Looks better, don't it?

File: 3600/3/helpers/comment_helper.rb [hunk 14]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+module CommentsHelper
+  def add_new_comment
+    @user = User.find(session[:user_id])
```
Comment (Moleved) [inline 3600/3/helpers/comment_helper.rb]: Could be a method of controller named `current_user`
```ruby
### app_controller.rb

def current_user
  @current_user ||= User.find_by(id: session[:user_id])
end
```

File: 3600/3/helpers/user_helper.rb [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  def valid_password?
+    return true if params['password'] == params['password_confirmation']
+
+    flash[:danger] = 'Passwords do not match'
+    redirect '/sign_up'
+  end
+
+  def valid_email?
```
Comment (Moleved) [inline 3600/3/helpers/user_helper.rb]: Check this https://guides.rubyonrails.org/active_record_validations.html#format

File: 3600/3/helpers/comment_helper.rb [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+    @comment = @user.comments.create(grade: params[:grade].to_i, text: params[:text])
+    @comment.restaurant_id = @restaurant.id
+    @comment.save
+  end
+
+  def logged_in?
```
Comment (Moleved) [inline 3600/3/helpers/comment_helper.rb]: Should be within `UserHelpers`

File: 3600/3/helpers/comment_helper.rb [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+    return true if session[:user_id]
+
+    flash[:danger] = 'You must be logged in!'
+    false
+  end
+
+  def negative_comment_has_text
```
Comment (Moleved) [inline 3600/3/helpers/comment_helper.rb]: Should be validated by `Comments` model not here

File: 3600/3/helpers/user_helper.rb [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+    flash[:danger] = 'No such user!'
+    redirect '/sign_in'
+  end
+
+  def sign_in_user
+    session[:user_id] = @user.id
+    redirect '/'
+  end
+
+  def valid_log_state
```
Comment (Moleved) [inline 3600/3/helpers/user_helper.rb]: `logged_in?` method is it you?

File: 3600/3/helpers/user_helper.rb [hunk 19]
```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+module UserHelper
+  def add_user
+    @user[:password] = Digest::SHA1.hexdigest(params[:password])
+    @user.save
+    session[:user_id] = @user.id
```
Comment (Moleved) [inline 3600/3/helpers/user_helper.rb]: `sign_in_user` method should be here

File: 3600/3/.gitignore (hunk 9 above)
Comment (sametsJR) [reply 3600/3/.gitignore]: It's my mistake too

File: 3600/3/app/controllers/user_controller.rb [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+  get '/sign_up' do
+    erb :sign_up
+  end
+
+  get '/sign_in' do
+    erb :sign_in
+  end
+
+  post '/sign_up' do
+    @user = User.new(name: params['username'], email: params['email'], password: params['password'])
```
Comment (Moleved) [inline 3600/3/app/controllers/user_controller.rb]:
```ruby
@user = User.create(..., password: encrypted_password)

private

def encrypted_password
  Digest::SHA1.hexdigest(params[:password])
end
```

File: 3600/3/helpers/comment_helper.rb (hunk 13 above)
Comment (sametsJR) [reply 3600/3/helpers/comment_helper.rb]: This line will be longer than 100 characters.

File: 3600/3/app/views/home.erb [hunk 21]
```diff
@@ -0,0 +1,52 @@
+<!DOCTYPE html>
```
Comment (Moleved) [inline 3600/3/app/views/home.erb]: Layouts are using for avoiding a lot of code within view files

File: 3600/3/app/views/home.erb [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+    <% @all_restaurants.each do |restaurant| %>
+      <div class="card mt-2 ml-2" style="width: 18rem;">
+        <div class="card-body">
+          <h5 class="card-title"> <%= restaurant.name %> </h5>
+          <p class="card-text"> <%= restaurant.short_description %> </p>
+          <a href="/restaurant/<%= restaurant.name %>" class="btn btn-outline-secondary">More info</a>
+        </div>
+      </div>
+    <% end %>
+  </div>
```
Comment (Moleved) [inline 3600/3/app/views/home.erb]: Leave in this file lines from 38 to 48 anything else move to layout

File: 3600/3/app/views/home.erb (hunk 21 above)
Comment (Moleved) [reply 3600/3/app/views/home.erb]: ERB is a preprocessor, so the file should be named like `my_awesom_view.html.erb`

File: 3600/3/app/controllers/restaurant_controller.rb [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+
+class RestaurantController < AppController
+  include CommentsHelper
+
+  get '/' do
+    @all_restaurants = Restaurant.all
+    erb :home
+  end
+
+  get '/restaurant/:name' do
```
Comment (Moleved) [inline 3600/3/app/controllers/restaurant_controller.rb]: Take a look at REST specification https://restfulapi.net/resource-naming/
According to it as a parameter should be provided `:id` not name

File: 3600/3/app/controllers/restaurant_controller.rb [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+    @all_comments = Comment.all
+    if @restaurant
+      erb :restaurant
+    else
+      flash[:danger] = 'No such cafe!'
+      redirect '/'
+    end
+  end
+
+  post '/restaurant/new-comment' do
```
Comment (Moleved) [inline 3600/3/app/controllers/restaurant_controller.rb]: This endpoint should be inside `CommentsController` which you don't have, so just name it right `/restaurant/:id/comment/`

File: 3600/3/app/controllers/restaurant_controller.rb [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+    @restaurant = Restaurant.find_by(name: session[:place])
```
Comment (Moleved) [inline 3600/3/app/controllers/restaurant_controller.rb]: According to the previous comment you don't need put place name into session just find it by id from params

File: 3600/3/app/controllers/restaurant_controller.rb [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+
+    add_new_comment if (negative_comment_has_text || params[:grade].to_i >= 3) && session[:user_id]
```
Comment (Moleved) [inline 3600/3/app/controllers/restaurant_controller.rb]: Check `validates :field, if: condition?` to simplify `(negative_comment_has_text || params[:grade].to_i >= 3)`

File: 3600/3/helpers/comment_helper.rb [hunk 27]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+# rubocop:disable Metrics/LineLength
+
+module CommentsHelper
+  def add_new_comment
+    @current_user.comments.create(grade: params[:grade].to_i, text: params[:text], restaurant_id: @restaurant.id)
+    @comment.restaurant_id = @restaurant.id
```
Comment (Moleved) [inline 3600/3/helpers/comment_helper.rb]: Now this line is useless

File: 3600/3/helpers/comment_helper.rb [hunk 28]
```diff
@@ -0,0 +1,12 @@
...
+    @comment.save
```
Comment (Moleved) [inline 3600/3/helpers/comment_helper.rb]: And this one too

File: 3600/3/helpers/comment_helper.rb [hunk 29]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+# rubocop:disable Metrics/LineLength
+
+module CommentsHelper
+  def add_new_comment
+    @current_user.comments.create(grade: params[:grade].to_i, text: params[:text], restaurant_id: @restaurant.id)
```
Comment (Moleved) [inline 3600/3/helpers/comment_helper.rb]: You may move this line into controller

Comment (Moleved) [reply 3600/3/helpers/comment_helper.rb]: Also `@current_user` available as a method call `current_user`

File: 3600/3/app/controllers/user_controller.rb (hunk 20 above)
Comment (Moleved) [reply 3600/3/app/controllers/user_controller.rb]: When I'm leaving such comments I'm expecting that I will see this code in your PR

File: 3600/3/helpers/user_helper.rb [hunk 30]
```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+# rubocop:disable Metrics/LineLength
+
+module UserHelper
+  def add_user
```
Comment (Moleved) [inline 3600/3/helpers/user_helper.rb]: Unnecessary method

File: 3600/3/app/controllers/restaurant_controller.rb [hunk 31]
```diff
@@ -0,0 +1,29 @@
...
+  include CommentsHelper
+
+  get '/' do
+    @all_restaurants = Restaurant.all
+    erb :home
+  end
+
+  get '/restaurant/:name' do
+    session[:place] = params[:name]
+    @restaurant = Restaurant.all.find_by(name: params[:name])
```
Comment (Moleved) [inline 3600/3/app/controllers/restaurant_controller.rb]: `find_by` working without `all`

File: 3600/3/app/controllers/restaurant_controller.rb [hunk 32]
```diff
@@ -0,0 +1,29 @@
...
+class RestaurantController < AppController
+  include CommentsHelper
+
+  get '/' do
+    @all_restaurants = Restaurant.all
+    erb :home
+  end
+
+  get '/restaurant/:name' do
+    session[:place] = params[:name]
```
Comment (Moleved) [inline 3600/3/app/controllers/restaurant_controller.rb]: When you will did what said in a comment for 23rd line this line would unnecessary

**Pull request merged**